edge-modules/api-proxy-module/src/monitors/token_client.rs

```diff
@@ -0,0 +1,142 @@
+use std::env;
+
+use anyhow::{Context, Error, Result};
+use chrono::{DateTime, Duration, Utc};
+use log::{error, info};
+use tokio::time;
+use url::form_urlencoded::Serializer as UrlSerializer;
+
+//Request new token 5min before expiry of actual token
+const TOKEN_EXPIRY_SECONDS_MARGIN: i64 = 300;
+
+pub fn get_token_client(validity_hours: Duration) -> Result<TokenClient, Error> {
+    let device_id =
+        env::var("IOTEDGE_DEVICEID").context(format!("Missing env var {}", "IOTEDGE_DEVICEID"))?;
+    let module_id =
+        env::var("IOTEDGE_MODULEID").context(format!("Missing env var {}", "IOTEDGE_MODULEID"))?;
+    let generation_id = env::var("IOTEDGE_MODULEGENERATIONID")
+        .context(format!("Missing env var {}", "IOTEDGE_MODULEGENERATIONID"))?;
+    let iothub_hostname = env::var("IOTEDGE_IOTHUBHOSTNAME")
+        .context(format!("Missing env var {}", "IOTEDGE_IOTHUBHOSTNAME"))?;
+    let workload_url = env::var("IOTEDGE_WORKLOADURI")
+        .context(format!("Missing env var {}", "IOTEDGE_WORKLOADURI"))?;
+
+    TokenClient::new(
+        device_id,
+        module_id,
+        generation_id,
+        iothub_hostname,
+        &workload_url,
+        validity_hours,
+    )
+}
+
+pub const IOTHUB_ENCODE_SET: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS // C0 control
+    .add(b' ')
+    .add(b'"')
+    .add(b'<')
+    .add(b'>')
+    .add(b'`') // fragment
+    .add(b'#')
+    .add(b'?')
+    .add(b'{')
+    .add(b'}') // path
+    .add(b'=');
+
+pub struct TokenClient {
+    device_id: String,
+    module_id: String,
+    generation_id: String,
+    iothub_hostname: String,
+    work_load_api_client: edgelet_client::WorkloadClient,
+    token_expiration_date: Option<DateTime<Utc>>,
+    validity_duration: Duration,
+}
+
+impl TokenClient {
+    pub fn new(
+        device_id: String,
+        module_id: String,
+        generation_id: String,
+        iothub_hostname: String,
+        workload_url: &str,
+        validity_duration: Duration,
+    ) -> Result<Self, Error> {
+        //Check duration more than margin
+
+        //Create expiry date in the past so cert has to be rotated now.
+        let token_expiration_date = None;
+
+        let work_load_api_client =
+            edgelet_client::workload(&workload_url).context("Could not get workload client")?;
+
+        Ok(TokenClient {
+            device_id,
+            module_id,
+            generation_id,
+            iothub_hostname,
+            work_load_api_client,
+            token_expiration_date,
+            validity_duration,
+        })
+    }
+
+    pub async fn get_new_sas_token(
+        &mut self,
+        current_date: DateTime<Utc>,
+    ) -> Result<Option<String>, Error> {
+        //If there is no previous token request a new one right away
+        if let Some(expiration_date) = self.token_expiration_date {
+            //If Token is not expired, sleep until it is.
+            use std::convert::TryInto;
```

Comment:
Can you pls explain what the logic here is doing?

Reply:
Let's chat about it.